File: tree.php
<?php
class Tree {
  private $dirCount;
  private $fileCount;

  public function __construct() {
    $this->dirCount = 0;
    $this->fileCount = 0;
  }

  public function summary() {
    return $this->dirCount . " directories, " . $this->fileCount . " files";
  }

  public function walk($directory, $prefix = "") {
    $filepaths = scandir($directory);

    foreach($filepaths as $index => $filepath) {
      if($filepath[0] == ".") {
        continue;
      }
      $absolute = join("/", array($directory, $filepath));
      $this->register($absolute);

      if($index == count($filepaths) - 1) {
        echo $prefix, "└── ", $filepath, "\n";
        if(is_dir($absolute)) {
          $this->walk($absolute, $prefix . "    ");
        }
      } else {
        echo $prefix, "├── ", $filepath, "\n";
        if(is_dir($absolute)) {
          $this->walk($absolute, $prefix . "│   ");
        }
      }
    }
  }

  private function register($filepath) {
    if(is_dir($filepath)) {
      $this->dirCount += 1;
    } else {
      $this->fileCount += 1;
    }
  }
}

$tree = new Tree();
$directory = ".";
if(count($argv) > 1) {
  $directory = $argv[1];
}

echo $directory, "\n";
$tree->walk($directory);
echo "\n", $tree->summary(), "\n";
?>

File: tree.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import sys

class Tree:
    def __init__(self):
        self.dirCount = 0
        self.fileCount = 0

    def register(self, absolute):
        if os.path.isdir(absolute):
            self.dirCount += 1
        else:
            self.fileCount += 1

    def summary(self):
        return str(self.dirCount) + " directories, " + str(self.fileCount) + " files"

    def walk(self, directory, prefix = ""):
        filepaths = sorted([filepath for filepath in os.listdir(directory)])

        for index in range(len(filepaths)):
            if filepaths[index][0] == ".":
                continue

            absolute = os.path.join(directory, filepaths[index])
            self.register(absolute)

            if index == len(filepaths) - 1:
                print(prefix + "└── " + filepaths[index])
                if os.path.isdir(absolute):
                    self.walk(absolute, prefix + "    ")
            else:
                print(prefix + "├── " + filepaths[index])
                if os.path.isdir(absolute):
                    self.walk(absolute, prefix + "│   ")

directory = "."
if len(sys.argv) > 1:
    directory = sys.argv[1]
print(directory)

tree = Tree()
tree.walk(directory)
print("\n" + tree.summary())
